File: Blind75/6. Most Common Elements/Most_Common_Elements.js
// Brute Force
function mostCommonElements(numbers, k){
    let countMap = new Map()

    for (const number of numbers){
        countMap.set(number, (countMap.get(number) || 0 ) + 1)
    }

    const sortedCountMap = [...countMap].sort((a, b) => b[1] - a[1]).splice(0, k)
    let result = []
    
    for (const num of sortedCountMap){
        result.push(num[0])
    }

    return result
}

// Optimized Approach using minPriorityQueue
function mostCommonElements(numbers, k){
    let countMap = new Map()

    for (const number of numbers){
        countMap.set(number, (countMap.get(number) || 0) + 1)
    }

    let pq = new myMinPriorityQueue()

    for(const [val, freq] of countMap){
        pq.add({val, freq})
        if (pq.size() > k){
            pq.remove()
        }
    }
    return pq.heap.map(item => item.val)
}

class myMinPriorityQueue{
    constructor(){
        this.heap = []
    }

    size(){
        return this.heap.length
    }

    getLeftElementIndex(node){
        return (node * 2) + 1
    }

    getRightElementIndex(node){
        return (node * 2) + 2
    }

    getParentElementIndex(node){
        return Math.floor((node - 1)/ 2)
    }

    add({val, freq}){
        this.heap.push({val, freq})
        this.heapifyUp()
    }

    heapifyUp(){
       let currNodeIndex = this.size() - 1

       while(currNodeIndex > 0){
            const parentElementIndex = this.getParentElementIndex(currNodeIndex)
            if(this.heap[currNodeIndex].freq < this.heap[parentElementIndex].freq){
                [this.heap[currNodeIndex], this.heap[parentElementIndex]] = [this.heap[parentElementIndex], this.heap[currNodeIndex]]
                currNodeIndex = parentElementIndex
            }
            else{
                break;
            }
       }
    }

    remove(){
        const lastNodeIndex = this.size() - 1;
        [this.heap[0], this.heap[lastNodeIndex]] = [this.heap[lastNodeIndex], this.heap[0]]
        this.heap.pop() 
        this.heapifyDown()
    }

    heapifyDown(){
        let currNodeIndex = 0
        let length = this.size()

        while(currNodeIndex < this.size()){
            const leftElementIndex = this.getLeftElementIndex(currNodeIndex)
            const rightElementIndex = this.getRightElementIndex(currNodeIndex)
            let smallestIndex = currNodeIndex

            if (leftElementIndex < length && this.heap[leftElementIndex].freq < this.heap[currNodeIndex].freq){
                smallestIndex = leftElementIndex
            }
            if(rightElementIndex < length && this.heap[rightElementIndex].freq < this.heap[smallestIndex].freq){
                smallestIndex = rightElementIndex
            }
            if(smallestIndex !== currNodeIndex){
                [this.heap[currNodeIndex], this.heap[smallestIndex]] = [this.heap[smallestIndex], this.heap[currNodeIndex]]
                currNodeIndex = smallestIndex
            }
            else{
                break;
            }
        }
    }
}

const testCases = [
    {"numbers": [4,4,4,6,6,5,5,5,9], "k": 2},
    {"numbers": [7,7,7,8,8,9,9,9], "k": 3},
    {"numbers": [10,10,10,10,10], "k": 1},
]

for (const {numbers, k} of testCases){
    console.log(mostCommonElements(numbers, k))
}

File: Blind75/Most Common Elements/Most_Common_Elements.py
def mostCommonElements(numbers: list[int], k: int) -> list[int]:
    count_map = {}
    
    for number in numbers:
        if number in count_map:
            count_map[number] += 1            
        else:
            count_map[number] = 1
    
    result = []
    sorted_array = sorted(count_map.items(), key=lambda item:item[1], reverse=True)[:k]
    
    for i in range(k):
        result.append(sorted_array[i][0])
    
    return result

# Optimized Approach 
def mostCommonElements(numbers: list[int], k: int) -> list[int]:
    count_map = {}
    
    for number in numbers:
        if number in count_map:
            count_map[number] += 1
        else: 
            count_map[number] = 1
            
    pq = MinPriorityQueue()
    
    for val, frequency in count_map.items():
        pq.append(val, frequency)
        if pq.size() > k:
            pq.remove()
    
    return [item[0] for item in pq.heap]

class MinPriorityQueue:
    def __init__(self):
        self.heap = []

    def size(self):
        return len(self.heap)
    
    def append(self, value, frequency):
        self.heap.append((value, frequency))
        self.heapifyUp()
        
    def getParentNodeIndex(self, node):
        return (node - 1 ) // 2
    
    def getLeftNodeIndex(self, node):
        return (node * 2) + 1
    
    def getRightNodeIndex(self, node):
        return (node * 2) + 2

    def heapifyUp(self):
        node_index =  len(self.heap) - 1
        
        while node_index > 0:
            parent_node_index = self.getParentNodeIndex(node_index)
            if self.heap[parent_node_index][1] > self.heap[node_index][1]:
                self.heap[parent_node_index], self.heap[node_index] = self.heap[node_index], self.heap[parent_node_index]
                node_index = parent_node_index
            else: 
                break
        
    def remove(self):
        last_node_index = len(self.heap) - 1
        self.heap[last_node_index], self.heap[0] = self.heap[0], self.heap[last_node_index]
        self.heap.pop()
        self.heapifyDown()
        
    def heapifyDown(self):
        node_index = 0
        length = len(self.heap)
        
        while True:
            left_node_index = self.getLeftNodeIndex(node_index)
            right_node_index = self.getRightNodeIndex(node_index)
            min_index = node_index
            
            if left_node_index < length and self.heap[left_node_index][1] < self.heap[min_index][1]:
                min_index = left_node_index

            if right_node_index < length and self.heap[right_node_index][1] < self.heap[min_index][1]:
                min_index = right_node_index
            
            if node_index != min_index:
                self.heap[node_index], self.heap[min_index] = self.heap[min_index], self.heap[node_index]
                node_index = min_index     
                
            else: 
                break                  
                
test_cases = [
    {"numbers": [4,4,4,6,6,5,5,5, 9], "k": 2},
    {"numbers": [7,7,7,8,8,9,9,9], "k": 3},
    {"numbers": [10,10,10,10,10], "k": 1},
]

for test_case in test_cases:
    print(mostCommonElements(**test_case))